### PR TITLE
Add / extend Cap'n Proto interfaces for key drivers.

### PR DESCRIPTION
**Pull request by kentonv**
Note that while the outgoing-HTTP and email drivers may be implemented as pseudo-drivers, I think that the web publishing driver may have no such luck, as it needs to store significant amounts of data and we will want to charge that against the user's quota. Plus we don't have a great place to store the data if not in a grain.

@zarvox @jparyani

### CODE REVIEW

File: src/sandstorm/web-publishing.capnp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  # under that URL. Note that the web driver assigns *no* special meaning to the character '/', so
+  # you should use a suffix ending in '/' if you want a site that represents a subdirectory.
+
+  # TODO(someday): Allow registering dynamaic handlers.
+
+  # TODO(someday): Allow transactionally changing a bunch of things at once.
+}
+
+interface WebResource {
+  # Represents an HTTP resource. Although StaticResource is an interface, an implementation must
```
Comment (dwrensha) [inline src/sandstorm/web-publishing.capnp]: I guess you renamed this from `StaticResource` to `WebResource` at some point?

File: src/sandstorm/email.capnp [hunk 2]
```diff
@@ -27,8 +27,17 @@ using Util = import "util.capnp";
 struct EmailAddress {
   # Email addresses are (usually) of the form "Full Name <example@example.org>".
   # This struct separates the display name and address into distinct fields.
+
   address @0 :Text;
   name @1 :Text;
+
+  # TODO(someday): We could add a field `token` which is a capability representing "email captial",
```
Comment (dwrensha) [inline src/sandstorm/email.capnp]: s/captial/capital/

File: src/sandstorm/web-publishing.capnp (hunk 1 above)
Comment (kentonv) [inline src/sandstorm/web-publishing.capnp]: Yikes, WebResource is an older version of my thoughts which I meant to delete entirely before pushing, but apparently forgot. Removed.

File: src/sandstorm/email.capnp (hunk 2 above)
Comment (kentonv) [inline src/sandstorm/email.capnp]: fixed

File: src/sandstorm/web-publishing.capnp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+# limitations under the License.
+
+@0xd5d3e63bd0a552b6;
+
+$import "/capnp/c++.capnp".namespace("sandstorm");
+
+using Util = import "util.capnp";
+using Handle = Util.Handle;
+
+interface WebSite {
```
Comment (zarvox) [inline src/sandstorm/web-publishing.capnp]: We should specify an explicit stable interface ID, if we want people to powerbox request it.

File: src/sandstorm/web-publishing.capnp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  # until enough data has actually been uploaded for them to respond. If the stream is dropped
+  # without calling `done()`, the blob will be left in a broken state in which some methods will
+  # throw exceptions.
+
+  getSubsite @4 (suffix :Text) -> (site :WebSite);
+  # Append `suffix` to the URL and return a `WebSite` capability that can only modify resources
+  # under that URL. Note that the web driver assigns *no* special meaning to the character '/', so
+  # you should use a suffix ending in '/' if you want a site that represents a subdirectory.
+
+  # TODO(someday): Allow registering dynamaic handlers.
```
Comment (zarvox) [inline src/sandstorm/web-publishing.capnp]: s/dynamaic/dynamic/

File: src/sandstorm/web-publishing.capnp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    body :union {
+      bytes @3 :Data;
+      # Use when content is smaller than 1MB.
+
+      blob @4 :Util.Blob;
+      # Use when content is larger than 1MB. Currently, the blob must be created via uploadBlob().
+    }
+  }
+
```
Comment (zarvox) [inline src/sandstorm/web-publishing.capnp]: Is there any way to enumerate the currently-known entities?  This API appears to require that you know a priori the exact full path of anything that you want to examine or modify, which may make removing content challenging.

File: src/sandstorm/web-publishing.capnp [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    mimeType @0 :Text;  # Content-Type header.
+    language @1 :Text;  # Content-Language header (optional).
+    encoding @2 :Text;  # Content-Encoding header (optional).
+
+    body :union {
+      bytes @3 :Data;
+      # Use when content is smaller than 1MB.
+
+      blob @4 :Util.Blob;
+      # Use when content is larger than 1MB. Currently, the blob must be created via uploadBlob().
```
Comment (dwrensha) [inline src/sandstorm/web-publishing.capnp]: I suppose the idea here is that someday this blob could alternatively come from an object-storage grain?

File: src/sandstorm/web-publishing.capnp [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    }
+  }
+
+  getEntities @1 (path :Text) -> (entities :Util.Assignable(List(Entity)));
+  # Get the list of static HTTP entities that make up the resource at the given path. Often there
+  # is only one entity, but if you provide multiple entities, one will be chosen based on "Accept"
+  # headers provided by the client. Setting the entity list empty effectively deletes the resource;
+  # the client will receive a 404 error, which can be customizedb using `getNotFoundEntities()`.
```
Comment (dwrensha) [inline src/sandstorm/web-publishing.capnp]: s/customizedb/customized/

File: src/sandstorm/api-session.capnp [hunk 8]
```diff
@@ -30,7 +30,103 @@ interface ApiSession @0xc879e379c625cdc7 extends(WebSession.WebSession) {
...
     # request.  This would be a privacy leak for WebSession, since the grain can give the client
-    # scripts which would send the header, but ApiSession requires a user action, so it's safe here.
+    # scripts which would send the header, but ApiSession requires a user action, so it's safe
+    # here.
     remoteAddress @0 :IpAddress;
   }
+
+  struct PowerboxTag {
+    # A tag used for PowerboxDescriptors describing HTTP APIs. Such descriptors should typically
+    # indicate `ApiSession` as the requested interface and use HttpApiTag to specify what API is
```
Comment (dwrensha) [inline src/sandstorm/api-session.capnp]: Should this `HttpApiTag` be defined somewhere?

File: src/sandstorm/api-session.capnp [hunk 9]
```diff
@@ -30,7 +30,103 @@ interface ApiSession @0xc879e379c625cdc7 extends(WebSession.WebSession) {
...
+    # expected.
```
Comment (dwrensha) [inline src/sandstorm/api-session.capnp]: So the ID of `PowerboxTag` is meant to be used in `PowerboxDescriptor.Tag.id`, with a value of type `PowerboxTag` in `PowerboxDescriptor.Tag.value`?

File: src/sandstorm/api-session.capnp (hunk 8 above)
Comment (dwrensha) [inline src/sandstorm/api-session.capnp]: This makes more sense if I s/HttpApiTag/PowerboxTag/.

File: src/sandstorm/api-session.capnp [hunk 10]
```diff
@@ -30,7 +30,103 @@ interface ApiSession @0xc879e379c625cdc7 extends(WebSession.WebSession) {
...
+    # driver will return a capability with a descriptor matching one of the requested descriptors
+    # which best-describes the user's choice.
+
+    canonicalUrl @0 :Text;
+    # The standard URL at which this service is found. Use this especially for traditional SaaS
+    # services. For example, you might request "https://api.github.com" to request access to the
+    # GitHub API.
+    #
+    # If you include a path in the URL, requests will automatically be prefixed with that path. You
+    # should usually include enough path components to identity a specific product and version, but
```
Comment (dwrensha) [inline src/sandstorm/api-session.capnp]: s/identity/identify/

**Comment by dwrensha** [comment]
I don't see anything in web-publishing.capnp about 300-level redirect responses. Do we want to support redirects, and how would they fit in with what you've designed here?

File: src/sandstorm/web-publishing.capnp [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+interface WebSite {
+  # Represents an HTTP URL to which a web site can be published.
+  #
+  # Make a powerbox request for `WebSite` when you wish to publish a web site accessible over HTTP
+  # without going through the Sandstorm UI.
+
+  getUrl @0 () -> (path :Text);
+  # Get the URL of this resource.
+
+  struct Entity {
```
Comment (zarvox) [inline src/sandstorm/web-publishing.capnp]: This interface should probably grow support for permitting cross-origin queries, so we can remove the hacks at https://github.com/sandstorm-io/sandstorm/blob/cf984260cca26a1d5b7abad82ee8fed73e5ff2d9/shell/server/pre-meteor.js#L86..L97

File: src/sandstorm/web-publishing.capnp (hunk 6 above)
Comment (kentonv) [inline src/sandstorm/web-publishing.capnp]: Yes. Or, raw Blackrock storage.

File: src/sandstorm/api-session.capnp (hunk 8 above)
Comment (kentonv) [inline src/sandstorm/api-session.capnp]: Yeah, I meant PowerboxTag. Went through some renames.

File: src/sandstorm/api-session.capnp (hunk 9 above)
Comment (kentonv) [inline src/sandstorm/api-session.capnp]: No, the type ID would be ApiSession's ID. The docs for PowerboxDescriptor specify that when Tag.id is an interface ID, Tag.value's type is specified by the interface's dogs. ApiSession specifies that ApiSession.PowerboxTag shall be used as the tag type.

File: src/sandstorm/web-publishing.capnp (hunk 11 above)
Comment (kentonv) [inline src/sandstorm/web-publishing.capnp]: Actually, I'd argue that this is best controlled not by the app or even the user, but by the server admin. If the system is on the public internet, then it's safe and appropriate to set `access-control-allow-origin: *` on all published sites. If the system is behind a firewall (only visible on some corp network), then `access-control-allow-origin` should specify only sites that are allowed to reach into the firewall. The app doesn't really have any reason to be expressing an opinion on the matter.

**Pull request merged**